### ISSUE TITLE
`pwnlib.misc.util.sh_string` broken for newline

### ISSUE THREAD

**Issue by zachriggle** [Gallopsled/pwntools#455]
If you pass *only* a newline to `sh_string`, the result is incorrect.  This may affect other input.

```py
>>> with open('\n','w+') as f: f.write("HELLO\n")
>>> exitcode = os.system("cat '\n'")
HELLO
>>> exitcode = os.system("cat %s" % sh_string('\n'))
cat: : No such file or directory
```

**Comment by zachriggle** [comment]
This appears to still be broken, but doesn't matter much because `ssh.process` is significantly more powerful, reducing reliance on `sh_string` with `ssh.run`.

**Comment by Idolf** [comment]
This is a problem with sh that I am not sure we can overcome.

Try doing the following two:

```sh
echo "$(echo a)"
echo "$(echo -n a)"
```

It seems that sh will always eat the trailing newline for convenience. Unless anybody has any good ideas, I think we should just add this as a limitation in the documentation.

**Comment by Idolf** [comment]
Alternatively we could just add an actual newline in the output, but we would also need to document this behavior.

**Comment by zachriggle** [comment]
However, the following works fine:

```
$ echo $'a\n'  | xxd
0000000: 610a 0a                                  a..
```

**Comment by Idolf** [comment]
Right, we can fix it in bash then, however that syntax does not work in a pure sh shell.

**Comment by zachriggle** [comment]
```
$ sh
sh-3.2$ echo $'a\n' | xxd
0000000: 610a 0a                                  a..
```

**Comment by Idolf** [comment]
Can you find me a quote on that being within the shell spec? It does not work in the openbsd shell.

**Comment by zachriggle** [comment]
It should work, but I don't have a mechanism to test OpenBSD.

OpenBSD docs don't mention it
<http://man.openbsd.org/cgi-bin/man.cgi?query=sh>.

FreeBSD docs say it should work
<https://www.freebsd.org/cgi/man.cgi?query=sh>.

     Dollar-Single Quotes
	     Enclosing characters between $' and ' preserves the literal mean-
	     ing of all	characters except backslashes and single quotes.  A
	     backslash introduces a C-style escape sequence:
...

	     \n		 Newline




*Zach Riggle*

**Comment by Idolf** [comment]
It seems that printf with octal escape sequences is the most portable option. It works on every machine that I have access to.

However I have still been unable to create a bash variable that contains a newline at then end without using $'\n'. I have tried the following

```sh
X1=$(printf '\012\012')
X2="$(printf '\012\012')"
X3=`printf '\012\012'`
X4="`printf '\012\012'`"
echo -n "$X1$X2$X3$X4" | xxd
```

Do you have any ideas?

**Comment by zachriggle** [comment]
I think with classic sh you can just embed the literal rather than escape
sequence

**Comment by Idolf** [comment]
From [stack overflow](https://stackoverflow.com/questions/15184358/how-to-avoid-bash-command-substitution-to-remove-the-newline-character), we should use something like this:

```sh
a=$(printf 'test\n\n'; printf x); a=${a%x}
```

It seems to be posix compliant.

**Comment by Idolf** [comment]
However this means that the goal of sh_string seems to be fundamentally impossible to do if the string contains trailing newlines.

**Comment by Idolf** [comment]
I have additionally added a new function called `sh_command_with`, which will use `sh_string` whenever that is good enough and use the above trick whenever it is not.

**Comment by Idolf** [comment]
Everything should be 100% posix compliant as far as I am aware.

**Comment by Idolf** [comment]
Closed in favor of discussing directly on #667.
